### DISCUSSION TITLE
any docs or examples for using nx for nodejs cli/terminal apps?

### DISCUSSION THREAD

**Discussion by firxworx** [nrwl/nx#13114] [General]
Are there any docs or examples for nodejs cli apps / terminal apps out there?

https://nx.dev/getting-started/node-tutorial teases because it creates an `apps/products-cli` however it is not in scope for the tutorial to actually complete it or provide more details.

Its not the first time I've been frustrated by nx docs and examples: when one topic is written about, examples get stubbed in but these are keyword traps to dead ends when looking for guidance on how to do the part that's stubbed in. My use-case is very similar to the node-tutorial except I would love a complete example!

Its super hard to find any resources because the "CLI" and "terminal" keyword space is cluttered by results about `nx` itself. 

In particular I'm curious what `project.json` should look like for executing a CLI app and importantly **accepting arguments**. 

If one uses the `@nrwl/js:node` executor it seems too limited and it actually ignores `process.exit()`.

What do you guys do for apps that run in the CLI terminal that you wrote in node?

Or do you just build them and use the published version in `dist/` with `nx:run-commands`... But then there's that args problem again!

**Comment by qortex** [comment]
I hit the same issue with the args. Did you find a good enough workflow?

For dev, I figured I could use the dev server without watch mode (works great it seems), but then args cannot be used (only static). Even after `--` in the command line.

Might do a PR to add a "forwardArgs" flag, it used to exist in some other executor I believe.

**Comment by firxworx** [comment]
@qortex I did end up finding a decent workflow after playing around with a few options: I don't run my cli through Nx :)

I have a few Nx projects now where I have written cli (I ended up choosing to create them in the workspace as lib/package vs. app for various reasons) and I just straight up run them using `tsx --tsconfig /path/to/my/libs/tsconfig.lib.json /path/to/my/libs/cli.ts` which I add that as a script in `package.json`.

I then run the cli via my package manager: `pnpm my_cli_script_name <args...>`

You will need to add [`tsx` ](https://www.npmjs.com/package/tsx) if you want to follow my approach: it will Just Work<sup>TM</sup> in a lot of cases where `ts-node` and others will cause headaches (e.g. if you import shared libraries/packages from within your Nx workspace in your cli code).

All the args get passed through no problems, and there's no chance of regression errors with Nx.

---

Your use-case sounds slightly different than mine given your description of watch mode. However you may find the general suggestion of taking Nx out of the equation where its causing a problem could lead you on a path that works for your project!
